### PR TITLE
Restrict inferred encodings for Uploads

### LINKED ISSUE #48945: CSV Imports Sometimes Incorrectly Infer Encoding (50.25+)
### Describe the bug

Imported a UTF8 CSV file with ASCII extended characters and it's importing them as Chinese characters. Interestingly whether this occurs seems to have some relationship to the size of the sample set in the file. the same row in the file 7 times will be imported correctly. But duplicate the 7 rows 10 more times and things go sideways.

The user initially experiencing this has large CSV files - oddly if the sample dataset is small the file imports correctly. Adding the same rows via copy/paste to the same file until the sample set is larger results in this bug presenting as well. So if you begin with a smaller CSV, then the encoding breaks as the files grows, you can no longer append to CSV, either.

### To Reproduce

1. Spin up version 50.25+ (works in 50.24)
2. Enabled CSV uploads for Postrges test DB
3. Import these test files:

- Incorrect: [import_test.csv](https://github.com/user-attachments/files/17464933/import_test.csv)
- Incorrect: [import_test2.csv](https://github.com/user-attachments/files/17464935/import_test2.csv)
- Correct: [import_test3.csv](https://github.com/user-attachments/files/17464938/import_test3.csv)

These import correctly in 50.24 and prior. They also work correctly if you use UTF8-BOM rather than UTF8.

### Expected behavior

Character encoding should be inferred correctly.

### Logs

_No response_

### Information about your Metabase installation

{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres"
    ],
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-09-10",
      "tag": "v1.50.25",
      "hash": "473a7ca"
    },
    "settings": {
      "report-timezone": null
    },
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "15.1 (Debian 15.1-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    }
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.24+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.24",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.24+8",
    "os.name": "Linux",
    "os.version": "5.15.153.1-microsoft-standard-WSL2",
    "user.language": "en",
    "user.timezone": "GMT"
  }
}

### Severity

blocking some users

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes #48945 

We've had some strange false positives for exotic encodings, see the linked issue.

As a conservative stopgap, we go back to assuming sane UTF-8 encoding in general, unless the encoding is well discernible and common. We can easily extend this list in future as requested by customers.

We'll want to backport this all the way to 50.

### CODE REVIEW

File: src/metabase/upload.clj [hunk 1]
```diff
@@ -282,25 +281,23 @@
...
+  #{"UTF-8"
+    "UTF-16" "UTF-16BE" "UTF-16LE"
+    "UTF-32" "UTF-32BE" "UTF-32LE"
+    "WINDOWS-1252"})
+
+(defn- detect-charset ^String [^File file]
+  (or
+   (try
+     ;; If its not a first-class supported encoding, just treat it as the default encoding.
+     (supported-charsets (UniversalDetector/detectCharset file))
```
Comment (crisptrutski) [inline src/metabase/upload.clj:294]: It turns out that they already had a helper for the tedious stream reading code. Basically the same as ours, with a smaller buffer.

File: test/metabase/upload_test.clj [hunk 2]
```diff
@@ -2171,18 +2182,19 @@
...
-                       {:upload-type bool-type,  :uncoerced "0.0",        :fail-msg "'0.0' is not a recognizable boolean"}
-                       {:upload-type int-type,   :uncoerced "01/01/2012", :fail-msg "'01/01/2012' is not a recognizable number"}]]
+                       #_{:upload-type int-type,   :uncoerced "2.0",        :coerced 2} ; value is coerced to int
+                       #_{:upload-type float-type, :uncoerced "2",          :coerced 2.0} ; column is promoted to float
+                       #_{:upload-type bool-type,  :uncoerced "0",          :coerced false}
+                       {:upload-type bool-type,  :uncoerced "2",          :coerced 2} ; column is promoted to int
+                       #_{:upload-type bool-type,  :uncoerced "1.0",        :fail-msg "'1.0' is not a recognizable boolean"}
+                       #_{:upload-type bool-type,  :uncoerced "0.0",        :fail-msg "'0.0' is not a recognizable boolean"}
+                       #_{:upload-type int-type,   :uncoerced "01/01/2012", :fail-msg "'01/01/2012' is not a recognizable number"}]]
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:2197]: Oops, this leaked from me playing with extensions to our promotion rules.

**Review by sansolovyov — APPROVED** [review]
+1 for restricting charsets (or at least heavily prioritizing utf-8)

is cp1252 heavily in use somewhere or what?

**Comment by crisptrutski** [comment]
It's fairly common, it's the default for Excel exports (or at least was for some time)

**Comment by crisptrutski** [comment]
@metabase-bot backport release-x.50.x

**Pull request merged**